sql/api/src/main/scala/org/apache/spark/sql/types/DataType.scala

```diff
@@ -408,6 +408,37 @@ object DataType {
     }
   }
 
+  /**
+   * Check if `from` is equal to `to` type except for collations and nullability, which are
+   * both checked to be compatible so that data of type `from` can be interpreted as of type `to`.
+   */
+  private[sql] def equalsIgnoreCompatibleCollationAndNullability(
+      from: DataType,
+      to: DataType): Boolean = {
+    (from, to) match {
+      case (_: StringType, _: StringType) => true
+
+      case (ArrayType(fromElement, fn), ArrayType(toElement, tn)) =>
+        (tn || !fn) && equalsIgnoreCompatibleCollationAndNullability(fromElement, toElement)
+
+      case (MapType(fromKey, fromValue, fn), MapType(toKey, toValue, tn)) =>
+        (tn || !fn) &&
+          // Map keys cannot change collation.
+          equalsIgnoreCompatibleNullability(fromKey, toKey) &&
+          equalsIgnoreCompatibleCollationAndNullability(fromValue, toValue)
+
+      case (StructType(fromFields), StructType(toFields)) =>
+        fromFields.length == toFields.length &&
```

Comment:
or do we use a map to not depend on the order of fields?

Reply:
Please check https://github.com/apache/spark/pull/46734#discussion_r1615598800.